sdk/src/main/java/io/radar/sdk/util/RadarFileSystemLogBuffer.kt

```diff
@@ -0,0 +1,212 @@
+package io.radar.sdk.util
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import io.radar.sdk.Radar
+import io.radar.sdk.model.RadarLog
+import org.json.JSONObject
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.util.*
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.locks.ReentrantLock
+
+/**
+ * Stores logs in files for persistent client-side logging
+ */
+internal class RadarFileSystemLogBuffer(
+    /**
+     * Application Context
+     */
+    context: Context,
+    /**
+     * Executor for writing files. This is used to ensure that log entries are written to the file in the order that
+     * they are received.
+     */
+    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
+) : RadarLogBuffer {
+
+    private companion object {
+
+        /**
+         * Name of the directory in which the logs are stored
+         */
+        const val DIRECTORY = "RadarLogs"
+
+        /**
+         * Maximum number of logs in the buffer.
+         */
+        const val LOG_BUFFER_SIZE_LIMIT = 500
+    }
+
+    /**
+     * The directory where the log files are stored
+     */
+    private val directory = context.getDir(DIRECTORY, Context.MODE_PRIVATE)
+
+    /**
+     * Lock for handling changes to the files within [DIRECTORY]
+     */
+    private val directoryLock = ReentrantLock()
+
+    /**
+     * Contains the name of the current log file. Modifications to this value should occur within a [fileOp]
+     * operation.
+     */
+    private var currentFile: String
+
+    /**
+     * Number of log entries in the current log file. Modifications to this value should occur within a [fileOp]
+     * operation.
+     */
+    var size: Int
+        private set
+
+    init {
+        // First find current file
+        val files = directory.listFiles()
+        if (files?.isNotEmpty() == true) {
+            // This gets the latest file that was in use if it can, otherwise it returns the first one.
+            currentFile = files.maxByOrNull { getLastModified(it) }!!.name
+            // This is initialized to the number of lines in the current file
+            size = FileInputStream(File(directory, currentFile)).bufferedReader().readLines().size
+            if (files.size > 2) {
+                // This case likely means the app started and ended several times without completing a purge. To prevent
+                // excessive log files, purge them now.
+                purgeInactiveFiles()
+            }
+        } else {
+            // If no such file exists yet, create one.
+            currentFile = getNewFilePath()
+            size = 0
+        }
+    }
+
+    override fun write(level: Radar.RadarLogLevel, message: String) {
+        executor.submit {
+            if (isFull()) {
```

Comment:
isFull takes a `fileOp` and looks like it's only used here, so I think ti might be worth rolling it all in here and wrapping this method in `fileOp` so there are less lock points to reason about

Reply:
now that I think about it. why bother with `fileOp` if we have `Executors.newSingleThreadExecutor()`?

can all methods be async / Future'd? this would mean there'd be no need to think about mutexes and locking, which are really hard to debug.

Reply:
I believe a lot of the end results/callers are things like publishing to the API server / asynchronous operations anyways, so it might fit this paradigm pretty well.

Reply:
We had this conversation in the previous PR - not all operations occur within the executor.

Reply:
This would also change how the logs are retrieved to send to the backend, and would require a callback instead of just returning it inline.